File: leetcode/Easy/validParenthesis.js
// Best solution for space O(n)

let isValid = function(s) {
    let stack = []
    if (s.length % 2 !== 0) {
        return false
    }
    for (let i = 0; i < s.length; i++) {
        let current = s.charAt(i)
        let topOfStack = stack[stack.length-1]
        if (current === "(" || current === "{" || current === "[") {
            stack.push(current)
        } else if (current === ")") {
            if (topOfStack === "(") {
                stack.pop()
            } else {
                return false
            }
        } else if (current === "]") {
            if (topOfStack === "[") {
                stack.pop()
            } else {
                return false
            }
        } else if (current === "}") {
            if (topOfStack === "{") {
                stack.pop()
            } else {
                return false
            }
        } else {
            return false
        }
    }
    if (stack.length !== 0) {
        return false
    }
    return true
};

// Below is the most optimal solution in terms of time and space. Specifically time.

let isValid2 = function(s) {
    let stack = []
    if (s.length % 2 !== 0) {
        return false
    }
    for (let i = 0; i < s.length; i++) {
        let current = s.charAt(i)
        if (current === "(" || current === "{" || current === "[") {
            stack.push(current)
        } else if (current === ")") {
            if (stack[stack.length-1] === "(") {
                stack.pop()
            } else {
                return false
            }
        } else if (current === "]") {
            if (stack[stack.length-1] === "[") {
                stack.pop()
            } else {
                return false
            }
        } else if (current === "}") {
            if (stack[stack.length-1] === "{") {
                stack.pop()
            } else {
                return false
            }
        } else {
            return false
        }
    }
    if (stack.length !== 0) {
        return false
    }
    return true
};

File: leetcode/Easy/validParenthesis.py
# Runtime - 90.92% and Memory - 13.66%

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        left = ["(", "[", "{"]
        right = [")", "]", "}"]
        for i in range(0, len(s)):
            if s[i] in left:
                stack.append(left.index(s[i]))
            if s[i] in right:
                if len(stack) == 0:
                    return False
                if stack[-1] == right.index(s[i]):
                    del stack[-1]
                else:
                    return False
        if len(stack) == 0:
            return True
        return False
            

# Did it again:
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        left = ["(", "[", "{"]
        right = [")", "]", "}"]

        for i in range(len(s)):
            if s[i] in left:
                stack.append(s[i])
            elif s[i] in right:
                if len(stack) == 0:
                    return False
                elif s[i] == ")":
                    if stack[-1] == "(":
                        stack.pop()
                    else:
                        return False
                elif s[i] == "]":
                    if stack[-1] == "[":
                        stack.pop()
                    else:
                        return False
                elif s[i] == "}":
                    if stack[-1] == "{":
                        stack.pop()
                    else:
                        return False
        if stack:
            return False
        return True
                